torch/csrc/jit/codegen/cuda/ir_nodes.cpp

```diff
@@ -1031,6 +1030,325 @@ std::pair<TensorDomain*, TensorDomain*> TensorDomain::rFactor(
+  // set_map implements the "parent" relationship
+  std::vector<int> set_map;
+  // weights is used for preliminary perf optimization
+  std::vector<int> weights;
+
+  // Map the input of type T to its equivalence class
+  std::unordered_map<T, int> entry_map;
+
+  // Utility for generating new class
+  int count = 0;
+
+  // internal fixed point implementation:
+  //  returns the equivalent class that e belongs to
+  //  does adhoc optimization to speed up future access
+  int fixedPoint(int e) {
+    TORCH_INTERNAL_ASSERT(set_map.size() > e);
+    while (set_map[e] != e) {
+      // optimization step
+      set_map[e] = set_map[set_map[e]];
+      // chasing to fixed point
+      e = set_map[e];
+    }
+    return e;
+  }
+
+  //! utility to check the class i belongs to:
+  //!
+  //! will create a new class if no match seen
+  //! \param i element i to find the equiv class for
+  //! \returns the equivalent class that e belongs to
+  //!
+  int fixedPoint(T i) {
+    // handles case when i doesn't have an equivalence class
+    if (!entry_map.count(i))
+      createPoint(i);
+    // use fixed point as a representation for the equiv class
+    return fixedPoint(entry_map[i]);
+  }
+
+  //! utility to create a new equiv class for i
+  //
+  //! \param i element i to create the equiv class for
+  void createPoint(T i) {
+    entry_map[i] = count;
+    set_map.push_back(count++);
+    weights.push_back(1);
+  }
+};
+
+//! Concretize broadcast axes, i.e. identifying a non-broadcast
+//! IterDomain that the broadcast IterDomain can map to.
+//!
+//! This traversal processes root domains only, concretization works by
+//! inspecting pointwise ops, e.g. : T2 [i0,i1] = T1[i0,B0] + T0[i0,i1]
+//! will concretize axis B0 to i1
+//!
+class ConcretizeDomain : public BackwardVisitor {
+ public:
+  //! traverses the graph backward from outputs
+  //! to identify all concretizing opportunities
+  //!
+  ConcretizeDomain(Fusion* fusion) {
+    traverseFrom(fusion, fusion->outputs(), false);
+  }
+
+  //! API call to run the concretize pass and return the
+  //! axis that bcast_dom concretizes to
+  //!
+  static const IterDomain* getConcreteDomain(IterDomain* bcast_dom) {
+    ConcretizeDomain cd(bcast_dom->fusion());
+
+    // remove this assertion once we support broadcast on output
+    TORCH_INTERNAL_ASSERT(cd.canConcretize(bcast_dom));
+    return cd.concretized(bcast_dom);
+  }
+
+  // returns true if either id is not a broadcast or
+  // the traversal has found a concretized axis for id
+  bool canConcretize(IterDomain* id) {
+    return !id->isBroadcast() || bcast_domain_map_.count(id);
+  }
+
+  // returns the concretized id recorded from traversal
+  IterDomain* concretized(IterDomain* id) {
+    TORCH_INTERNAL_ASSERT(canConcretize(id));
+    if (!id->isBroadcast())
+      return id;
+    return bcast_domain_map_.at(id);
+  }
+
+ private:
+  using MapType = std::unordered_map<IterDomain*, IterDomain*>;
+  MapType bcast_domain_map_;
```

Comment:
same comment regarding the separation of private state from private methods

Reply:
updated. Thanks!